elf/arch-riscv.cc

```diff
@@ -459,6 +463,70 @@ void InputSection<E>::apply_reloc_alloc(Context<E> &ctx, u8 *base) {
         set_rs1(loc, 4);
       break;
     }
+    case R_RISCV_TLSDESC_HI20: {
+      if (removed_bytes == 0) {
+        u64 desc = sym.get_tlsdesc_addr(ctx);
+        write_utype(loc, desc + A - P);
+      }
+      break;
+    }
+    case R_RISCV_TLSDESC_LOAD_LO12: {
+      i64 idx2 = find_paired_reloc();
+      const ElfRel<E> &rel2 = rels[idx2];
+      Symbol<E> &sym2 = *file.symbols[rel2.r_sym];
+      u64 A = rel2.r_addend;
+      u64 P = get_addr() + rel2.r_offset - get_r_delta(idx2);
+      if (sym2.has_tlsdesc(ctx)) {
+        u64 desc = sym2.get_tlsdesc_addr(ctx);
+        write_itype(loc, desc + A - P);
+      }
+      break;
+    }
+    case R_RISCV_TLSDESC_ADD_LO12: {
+      i64 idx2 = find_paired_reloc();
+      const ElfRel<E> &rel2 = rels[idx2];
+      Symbol<E> &sym2 = *file.symbols[rel2.r_sym];
+      u64 S = sym2.get_addr(ctx);
+      u64 A = rel2.r_addend;
+      u64 P = get_addr() + rel2.r_offset - get_r_delta(idx2);
+      if (sym2.has_tlsdesc(ctx)) {
+        u64 desc = sym2.get_tlsdesc_addr(ctx);
+        write_itype(loc, desc + A - P);
+      } else if (sym2.has_gottp(ctx)) {
+        u64 desc = sym2.get_gottp_addr(ctx);
+        *(ul32 *)loc = 0x517; // auipc a0,<hi20>
+        write_utype(loc, desc + A - P);
+      } else {
+        if (removed_bytes == 0) {
+          *(ul32 *)loc = 0x537; // lui a0,<hi20>
```

Comment:
Not directly related to this change, but I think there's one more relaxation opportunity here. If the hi20 bits are small enough, you can use C.LUI instead of LUI if the C extension is available, which should save two bytes.